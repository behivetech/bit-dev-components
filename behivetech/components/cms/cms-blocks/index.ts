export { CmsBlocks } from './cms-blocks';
export type { CmsBlocksProps } from './cms-blocks';
