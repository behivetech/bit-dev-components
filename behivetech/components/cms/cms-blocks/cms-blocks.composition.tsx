import * as React from 'react';
import { CmsBlocks } from './cms-blocks';
import {CmsProvider, useCmsProvider} from '@behivetech/behivetech.components.cms.cms-provider';

const MOCK_AREA_NAME = 'cmsBlocksComposition';

const MockImportedComponent = ({componentIndex}) => <div>Imported Block {componentIndex}</div>
const MOCK_IMPORTED_BLOCKS = {
    MockImportedComponent1: MockImportedComponent,
    MockImportedComponent2: MockImportedComponent,
    MockImportedComponent3: MockImportedComponent,
    MockImportedComponent4: MockImportedComponent,
};
const MOCK_CMS_PARAMS = {
    areas: {
        [MOCK_AREA_NAME]: {blocks: [
            {componentName: 'MockImportedComponent1', componentProps: {componentIndex: 1}},
            {componentName: 'MockImportedComponent2', componentProps: {componentIndex: 2}},
            {componentName: 'MockImportedComponent3', componentProps: {componentIndex: 3}},
            {componentName: 'MockImportedComponent4', componentProps: {componentIndex: 4}},
            {componentName: 'MockImportedComponent5', componentProps: {componentIndex: 5}}, // Should be a missing block
        ]}
    }
};

export const BasicCmsBlocks = () => (
    <CmsProvider cmsParams={MOCK_CMS_PARAMS}>
        <CmsBlocks areaName={MOCK_AREA_NAME} importedBlocks={MOCK_IMPORTED_BLOCKS} />
    </CmsProvider>
);
