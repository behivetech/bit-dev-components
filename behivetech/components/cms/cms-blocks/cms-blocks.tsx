import * as React from 'react';

import {CmsBlockMissing} from '@behivetech/behivetech.components.cms.cms-block-missing';
import {BlocksType, useCmsProvider} from '@behivetech/behivetech.components.cms.cms-provider';
import {ImportedBlock, ImportedBlocks} from '@behivetech/behivetech.components.cms.cms-area'

export type CmsBlocksProps = {
    /**
     * areaName is the key used to pull from the cmsParams from the context of the `<CmsProvider />`
     * component.
     */
    areaName: string,
    /**
     * an object of imported componets where the key is the same as the component name. The importedBlocks
     * should be an object of components specific to is pulled from the context cmsParams.
     */
    importedBlocks: ImportedBlocks
};

export function CmsBlocks({ areaName, importedBlocks }: CmsBlocksProps) {
    const {editBlock, cmsParams: {areas}} = useCmsProvider();
    const blocks = areas[areaName] ? areas[areaName]?.blocks : [];

    return (<>
        {
            blocks.map(({componentName, componentProps}, index) => {
                const componentKey = `${areaName}__${componentName}--${index}`;
                const DynamicBlock: ImportedBlock = importedBlocks[componentName];

                return DynamicBlock
                    ? <DynamicBlock key={componentKey} {...componentProps}/>
                    : <CmsBlockMissing componentName={componentName} key={componentKey} />
            })
        }
    </>);
}
