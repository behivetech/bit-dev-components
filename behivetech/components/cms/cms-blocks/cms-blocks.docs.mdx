---
description: 'A CmsBlocks component.'
labels: ['CMS', 'blocks']
---

import { CmsBlocks } from './cms-blocks';

## React Component for rendering CMS Blocks
Blocks are custom components created to display whatever is needed in the UI. These can be any matter of things such as a simple header or footer to display what is necessary for that part of the UI layout or they can be full micro-frontend applicaitons to diplay data coming from an API. The information coming from the cmsParams prop passed into the CmsProvider determines what blocks will be displayed within a CmsArea component. This component will never be used directly by a developer unless a custom component is created to replace the functionality of the `<CmsArea ... />` component. For more details, check out the documents provided in the `<CmsProvider />` component.

