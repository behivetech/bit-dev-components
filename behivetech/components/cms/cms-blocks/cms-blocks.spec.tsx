import * as React from 'react';
import { render } from '@testing-library/react';
import { BasicCmsBlocks } from './cms-blocks.composition';

it('should render with the correct content', () => {
    const { asFragment } = render(<BasicCmsBlocks />);

    expect(asFragment()).toMatchSnapshot();
});
