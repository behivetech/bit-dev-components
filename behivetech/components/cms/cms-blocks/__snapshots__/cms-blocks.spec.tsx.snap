// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render with the correct content 1`] = `
<DocumentFragment>
  <div>
    Imported Block 1
  </div>
  <div>
    Imported Block 2
  </div>
  <div>
    Imported Block 3
  </div>
  <div>
    Imported Block 4
  </div>
  <div
    class="cms-block-missing"
    style="background-color: rgb(153, 0, 0); color: rgb(255, 255, 255);"
  >
    COMPONENT IMPORT ERROR: The MockImportedComponent5 Componpent is not available from the importedBlocks prop.
  </div>
</DocumentFragment>
`;
